### PR TITLE
Add combined pyode and libode

### PR DESCRIPTION
**Pull request by repagh**
Checklist

- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example)
- [x] Source is from official source
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged)
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details)
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there

### CODE REVIEW

**Comment by repagh** [comment]
Puzzles to solve:
- linux fails because "import ode" does not find clock_gettime, which may be in librt
- windows build does not find ode.pc with pkg-config

**Comment by repagh** [comment]
This builds fine, except for the AppVeyor check. 

I cannot find anything in the AppVeyor log that points to a cause, and with the module working for me all three platforms, I don't see a reason that this check would point to a valid failure.

Can this be merged without passing AppVeyor, or is there someone who can give a pointer on fixing that test?

File: recipes/pyode/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+  number: 0
+  skip: true                         # [win and py==27]
+  
+requirements:
+  build:
+    - {{ compiler('cxx') }}          # stage 1, library
+    - cmake
+    - {{ compiler('c') }}            # stage 2, generate bindings
+    - pkg-config
+    - m2-patch                       # [win]
```
Comment (chrisburr) [inline recipes/pyode/meta.yaml]:
```suggestion
    - m2-patch                       # [win]
    - make                           # [unix]
```

File: recipes/pyode/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    - python
+
+  run:
+    - python
+
+test:
+  requires:
+    - python
+  commands:
+    - test -f $PREFIX/lib/libode.so                       # [linux]
```
Comment (chrisburr) [inline recipes/pyode/meta.yaml]: These 4 lines can be simplified a little using `${SHLIB_EXT}`:
```yaml
    - test -f $PREFIX/lib/libode${SHLIB_EXT}                       # [unix]
    - test -f $PREFIX/lib/libode${SHLIB_EXT}.{{ soversion }}       # [unix]
```

File: recipes/pyode/build.sh [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+      -DCMAKE_INSTALL_LIBDIR="lib"
+
+make VERBOSE=1
+make install DESTDIR="${PREFIX}"
+cd ..
+
+# stage 2, bindings
+
+# make sure recently installed ode.pc is found
+PKG_CONFIG_PATH=${PREFIX}/lib/pkgconfig:
```
Comment (chrisburr) [inline recipes/pyode/build.sh]: Is this needed? Naively, I expect it to not to do anything

File: recipes/pyode/LICENSE-BSD.TXT [hunk 4]
```diff
@@ -0,0 +1,34 @@
+
```
Comment (chrisburr) [inline recipes/pyode/LICENSE-BSD.TXT]: This file isn't needed as it's included in the source tarball and the `licence_file` field looks there automatically. It's also preferable as it ensures the file stays up to date if the licence ever changes.

File: recipes/pyode/LICENSE.TXT [hunk 5]
```diff
@@ -0,0 +1,502 @@
+		  GNU LESSER GENERAL PUBLIC LICENSE
```
Comment (chrisburr) [inline recipes/pyode/LICENSE.TXT]: This file isn't needed as it's included in the source tarball and the `licence_file` field looks there automatically. It's also preferable as it ensures the file stays up to date if the licence ever changes.

File: recipes/pyode/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+  host:
+    - cython
+    - python
+
+  run:
+    - python
+
+test:
+  requires:
```
Comment (chrisburr) [inline recipes/pyode/meta.yaml]: No need for this `requires` section as `python` is already in the `run` requirements.

File: recipes/pyode/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,63 @@
+# defining package and version
+{% set name = "pyode" %}
+{% set version = "0.16" %}
+{% set soversion = "0.16.0" %}
```
Comment (chrisburr) [inline recipes/pyode/meta.yaml]: Does this need to be a separate variable? Or can `{{ version }}.0` be used instead?

File: recipes/pyode/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+{% set sha256 = "4ba3b76f9c1314160de483b3db92b0569242a07452cbb25b368e75deb3cabf27" %}
+
+package:
+  name: {{ name|lower }}
+  version: {{ version }}
+
+source:
+  url: https://bitbucket.org/odedevs/ode/downloads/ode-{{ version }}.tar.gz
+  sha256: {{ sha256 }}
```
Comment (chrisburr) [inline recipes/pyode/meta.yaml]: As it's only used in one place I generally prefer not having a `sha256` variable and just writing it directly
```suggestion
  sha256: 4ba3b76f9c1314160de483b3db92b0569242a07452cbb25b368e75deb3cabf27
```

File: recipes/pyode/LICENSE-BSD.TXT (hunk 4 above)
Comment (repagh) [reply recipes/pyode/LICENSE-BSD.TXT]: ok, removing

File: recipes/pyode/LICENSE.TXT (hunk 5 above)
Comment (repagh) [reply recipes/pyode/LICENSE.TXT]: ok, removing

File: recipes/pyode/meta.yaml (hunk 7 above)
Comment (repagh) [reply recipes/pyode/meta.yaml]: don't know the plans, I would like to keep these uncoupled.

File: recipes/pyode/meta.yaml (hunk 2 above)
Comment (repagh) [reply recipes/pyode/meta.yaml]: simplified as suggested

File: recipes/pyode/build.sh (hunk 3 above)
Comment (repagh) [reply recipes/pyode/build.sh]: Indeed, not needed.

File: recipes/pyode/meta.yaml (hunk 6 above)
Comment (repagh) [reply recipes/pyode/meta.yaml]: ok, removed

**Comment by repagh** [comment]
> Looks good.
> 
> Would it make sense to have `libode` as a separate package that is built at the same time so it's possible to use without depending on Python?

I started out like that, but reverted, because I did not want to wait for the lib to get through until I could work on the binding. 

Is it possible to generate 2 packages from one recipe, like rpm or deb does? The python binding is included in the same source. 

On the other hand, anyone installing something from conda-forge will be using Python anyhow?

And what is the issue with the appveyor build. Does that simply need a general update in staged-recipes?

File: recipes/pyode/meta.yaml (hunk 2 above)
Comment (repagh) [reply recipes/pyode/meta.yaml]: You fooled me there. They can be simplified to 3, but look more complex...

**Comment by repagh** [comment]
Can this be merged? Or are there still issues to be resolved?

**Comment by chrisburr** [comment]
> Is it possible to generate 2 packages from one recipe, like rpm or deb does? The python binding is included in the same source.

Yes, see [here](https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#outputs-section) or [matplotlib](https://github.com/conda-forge/matplotlib-feedstock/blob/master/recipe/meta.yaml) for an example. If you want to add multiple outputs you can do it here. If not let me know and I'll merge.

> On the other hand, anyone installing something from conda-forge will be using Python anyhow?

There definitely are, conda isn't Python specific at all and many people use it for R, Java, C++ and others.

> And what is the issue with the appveyor build. Does that simply need a general update in staged-recipes?

I think Appveyor has been fixed now but regardless you can ignore it if it fails, the Azure check is enough.

File: recipes/pyode/meta.yaml [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+        - libode
+    files:
+        - lib/python*/site-packages/*
+    test:
+      imports:
+        - ode
+
+about:
+  home: https://www.ode.org/
+  license: LGPL-2.0 or BSD
```
Comment (chrisburr) [inline recipes/pyode/meta.yaml]: Which BSD licence is it? (please check I got it right, I only briefly looked)
```suggestion
  license: LGPL-2.0 or BSD-4-Clause
```

Comment (repagh) [reply recipes/pyode/meta.yaml]: Counting 4 clauses, so yes.

**Comment by repagh** [comment]
I'm sorta stuck here. 
`  ERROR (pyode,lib/python3.7/site-packages/ode.cpython-37m-x86_64-linux-gnu.so): did not find - or even know where to look for: /lib64/libc.so.6
  ERROR (pyode,lib/python3.7/site-packages/ode.cpython-37m-x86_64-linux-gnu.so): did not find - or even know where to look for: /lib64/librt.so.1
  ERROR (pyode,lib/python3.7/site-packages/ode.cpython-37m-x86_64-linux-gnu.so): lib/libode.so.0.16.0 not found in any packages
  ERROR (pyode,lib/python3.7/site-packages/ode.cpython-37m-x86_64-linux-gnu.so): did not find - or even know where to look for: /lib64/libpthread.so.0
`
On a local build, this results in warnings, and simply produces a working library and python module, that, I suspect, use the commonly installed librt, libpthread and libc.
Are there any hints as to how the conda packages are named for these libraries? Or do I need another approach?

**Comment by repagh** [comment]
I added a patch for Windows version, fixing an issue caused by trying to store 64 bit pointers in 32 bit integer (apparently "long" on 64-bit Windows is just 32 bit). This would not crash on import, only on use of a model. I tested with an ODE model on all three platforms, and all works. 

AppVeyor conked out again, but this seems unrelated to the package. 

Can this please be merged?

**Comment by repagh** [comment]
Discussed the patch on https://bitbucket.org/odedevs/ODE/issues
Preference is for a variable that would remain 4byte on 32-bit architectures, therefore using size_t instead of "long long", tested, works.

File: .appveyor.yml [hunk 10]
```diff
@@ -40,7 +40,7 @@ install:
     - cmd: appveyor-retry conda.exe update --yes --quiet conda
 
 
-    - cmd: appveyor-retry conda.exe install --yes --quiet "conda!=4.6.1,<4.7.11a0" conda-forge-pinning conda-forge-ci-setup=2.* networkx=2.3 "conda-build>=3.16"
+    - cmd: appveyor-retry conda.exe install --yes --quiet "conda!=4.6.1,<4.7.11a0" conda-forge-pinning conda-forge-ci-setup=2.* networkx=2.3 "conda-build>=3.18"
```
Comment (scopatz) [inline .appveyor.yml]: Please do not change files outside of the `recipes/` dir in recipe PRs. Thanks!

File: recipes/pyode/build.sh [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+      -DCMAKE_INSTALL_LIBDIR="lib"
+
+make VERBOSE=1
+make install DESTDIR="${PREFIX}"
+cd ..
+
+# stage 2, bindings
+
+# code has aliasing warnings, use flag to prevent crashes
+CFLAGS="-fno-strict-aliasing $CFLAGS"
```
Comment (scopatz) [inline recipes/pyode/build.sh]: This has no effect unless you export it

**Comment by scopatz** [comment]
A couple of more comments, thanks @repagh!

File: .appveyor.yml (hunk 10 above)
Comment (repagh) [reply .appveyor.yml]: reverted

File: recipes/pyode/build.sh (hunk 11 above)
Comment (repagh) [reply recipes/pyode/build.sh]: added the export

**Comment by scopatz** [comment]
Thanks!

**Pull request merged**